docker-app/qfieldcloud/core/adapters.py

```diff
@@ -69,3 +116,33 @@ def send_confirmation_mail(
             )
 
         super().send_confirmation_mail(request, email_confirmation, signup)
+
+
+class SocialAccountAdapter(DefaultSocialAccountAdapter):
+    """Custom SocialAccountAdapter to aid SSO integration in QFC.
+
+    Logs stack trace and error details on 3rd party authentication errors.
+    """
+
+    def on_authentication_error(
+        self,
+        request: HttpRequest,
+        provider: OAuth2Provider,
+        error: str = None,
+        exception: Exception = None,
+        extra_context: dict = None,
+    ) -> None:
+        logger.error("SSO Authentication error:", exc_info=True)
+        logger.error(f"Provider: {provider!r}")
+        logger.error(f"Error: {error!r}")
+
+        if not extra_context:
+            extra_context = {}
+
+        # Make stack strace available in template context.
+        extra_context["formatted_exception"] = "\n".join(
+            traceback.format_exception(exception)
+        )
```

Comment:
The only thing I am missing and probably it's worth adding a comment is which "template context" we are talking about here? I am sure if I deep dive in the docs I can find it, just acting lazy. If screenshot is faster in explaining this, feel free.

Reply:
This just means that it would be available as one of the template variables in the top level namespace. So in the `socialaccount/authentication_error.htm` template, this could then be used as `{{ formatted_exception }}`.

_(The superclass implementation of `on_authentication_error()`, called right below, will stuff everything from `extra_context` into the `context` that is passed to the template)._

I have added a comment in the code to explain this a bit more. 👍